controllers/community.server.controller.ts

```diff
@@ -0,0 +1,151 @@
+import * as User from '../models/user.server.model';
+import config from '../config/config.server.config';
+import { validate } from '../lib/validate.lib';
+import { PostResponse } from './forum.server.controller';
+
+interface UpdateCommunityDTO {
+  name?: string;
+  description?: string;
+  img?: string;
+}
+
+interface CreateCommunityDTO {
+  name: string;
+  description: string;
+  img?: string;
+}
+
+interface CommunityResponse extends ICommunity {
+  id: mongoose.Types.ObjectId;
+  memberCount: number;
+}
+
+export async function getPosts(req: Request, res: Response<PostResponse[]>) {
+  const id = convertToObjectId(req.params.id);
+  const posts = await Forum.populatePosts(await Community.getPosts(id));
+
+  const response = posts.map(
+    (post) =>
+      ({
+        id: post._id,
+        owner: post.owner,
+        community: {
+          id: post.community._id,
+          name: post.community.name,
+        },
+        title: post.title,
+        bodyText: post.bodyText,
+        edited: post.edited,
+        upVotes: post.upVotes,
+        attachments: post.attachments,
+        downVotes: post.downVotes,
+        comments: post.comments,
+        createdAt: post.createdAt,
+        updatedAt: post.updatedAt,
+      } as PostResponse),
+  );
+
+  res.status(200).send(response);
+}
+
+export async function communityUpdateById(
+  req: TypedRequestBody<UpdateCommunityDTO>,
+  res: Response<CommunityResponse>,
+) {
+  const authToken = req.get(config.get('authToken'));
+
+  const schema = Joi.object<UpdateCommunityDTO>({
+    name: Joi.string().min(3),
+    description: Joi.string().allow(''),
+    img: Joi.string().uri(),
+  }).min(1);
+
+  const data = validate(schema, req.body);
+
+  const id = convertToObjectId(req.params.id);
+  const community = await Community.searchCommunityById(id);
+
+  if (!(await User.isUserAuthorized(community.owner, authToken))) {
+    throw new ServerError('forbidden', 403);
+  }
+
+  const newCommunity = await Community.updateCommunityById(id, data);
+
+  res.status(200).send({
+    id: newCommunity._id,
+    img: newCommunity.img,
+    name: newCommunity.name,
+    owner: newCommunity.owner,
+    memberCount: await Community.getCommunityMemberCount(id),
+    description: newCommunity.description,
+    createdAt: newCommunity.createdAt,
+    updatedAt: newCommunity.updatedAt,
+  });
+}
+
+export async function getCommunities(
+  req: Request,
+  res: Response<CommunityResponse[]>,
+) {
+  const communities = await Community.getAll();
+
+  //Todo: get member counts asynchronously
+  const mappedCommunities: CommunityResponse[] = [];
+  for (const i of communities) {
+    mappedCommunities.push({
+      id: i._id,
+      img: i.img,
+      name: i.name,
+      owner: i.owner,
+      memberCount: await Community.getCommunityMemberCount(i._id),
+      description: i.description,
+      createdAt: i.createdAt,
+      updatedAt: i.updatedAt,
+    });
+  }
```

Comment:
Can this be changed to use map() instead? I think you can do await Promise.all() to solve the asynchronous part

Reply:
If I change it to map then I can't use await in the body of the map. I could solve that by getting the member counts first but I'm not sure how it would be best to do that

Reply:
https://flaviocopes.com/javascript-async-await-array-map/

Reply:
nice, implemented